### PR TITLE
[AWS] Add LB Type in the infrastructure cluster object via install-config yaml

### PR DESCRIPTION
**Pull request by miheer**
In the ingress config cluster object we need to add api for AwsLBType which will store and persist the information of the desired LB type mentioned by the installer config API so that the ingress operator will refer the LB type from this object whenever someone accidentally deletes the ingress controller.

Currently, the default LB type i.e CLB/ELB gets set if you delete the ingress controller object even when the desired the LB type for default ingress controller
was set to NLB as ingress operator does not have a way to check the LB type as we don't store it.
This commit adds LB Type for AWS mentioned in the install-config.yaml in the infrastructure cluster object.

`data/data/install.openshift.io_installconfigs.yaml`  A generated new field called lbType will be available to add load balancer type in the install-config.yaml for AWS
`pkg/asset/manifests/ingress.go`  The ingress cluster object will be patched with this lbType provided by user so that we have the information related to the lbType persistent which can be referred by ingress operator  when the default ingress controller is deleted.
`pkg/asset/manifests/ingress_test.go` Unit test to test setting of lbType in the ingress cluster object's spec and status as per the value specified in the install-config.yaml
`pkg/explain/printer_test.go` A new field to test called lbType which will be available to add the load balancer type in the install-config.yaml for AWS
`pkg/types/aws/platform.go` A new field called lbType will be available to add the load balancer type in the install-config.yaml for AWS

Enhacement proposal - https://github.com/openshift/enhancements/pull/1148
OpenShift API - https://github.com/openshift/api/pull/1209

Jira ticket - https://issues.redhat.com/browse/NE-942

### CODE REVIEW

**Comment by patrickdillon** [comment]
Please separate vendoring into a separate commit to make reviewing easier.

It would also be good to separate the pr into logical, smaller commits per the [contributing guidelines](https://github.com/openshift/installer/blob/master/CONTRIBUTING.md).

Are there any reference links for this work such as an enhancement or jira cards?

**Comment by miheer** [comment]
@patrickdillon thanks! I have done the changes you asked for. Please do let me know if anything is needed.

I have added the commit message in detail as to which file does what.

Also reference to enhancement proposal, API and jira ticket have been added in the commit message.

I am adding them here as well -

https://github.com/openshift/enhancements/pull/1148

https://github.com/openshift/api/pull/1209

**Comment by sadasu** [comment]
/hold
Enhacement proposal - https://github.com/openshift/enhancements/pull/1148
OpenShift API - https://github.com/openshift/api/pull/1209 
need to merge first.

File: pkg/asset/manifests/ingress.go [hunk 1]
```diff
@@ -108,13 +108,24 @@ func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte,
...
+			LoadBalancer: configv1.LoadBalancer{
+				Platform: configv1.IngressPlatformSpec{
+					AWS: &configv1.AWSIngressSpec{
+						Type: config.AWS.LBType,
+					},
+					Type: configv1.AWSPlatformType,
+				},
+			},
+		}
+	}
```
Comment (sadasu) [inline pkg/asset/manifests/ingress.go:128]: when LBType is not provided in install-config, do we want to set it to "Classic" here?

File: pkg/asset/manifests/ingress_test.go [hunk 2]
```diff
@@ -91,6 +93,15 @@ func TestGenerateIngerssDefaultPlacement(t *testing.T) {
...
 		},
+		{
+			name:                        "test setting of aws lb type",
+			installConfigBuildOptions:   []icOption{icBuild.withLBType(configv1.NLB)},
+			controlPlaneTopology:        configv1.HighlyAvailableTopologyMode,
+			infrastructureTopology:      configv1.HighlyAvailableTopologyMode,
+			expectedIngressPlacement:    configv1.DefaultPlacementWorkers,
+			expectedIngressAWSLBType:    configv1.NLB,
+			expectedIngressPlatformType: configv1.AWSPlatformType,
+		},
```
Comment (sadasu) [inline pkg/asset/manifests/ingress_test.go:104]: I think "vanilla aws" should be augmented to check if the expectedIngressAWSLBType and expectedIngressPlatformType continue to behave as expected for the default "Classic" case after these changes.

**Comment by sadasu** [comment]
/retitle [AWS] Add LB Type in the infrastructure cluster object via install-config yaml

File: pkg/asset/manifests/ingress.go (hunk 1 above)
Comment (miheer) [reply pkg/asset/manifests/ingress.go:128]: If no value is set then ingress operator will create a CR for Classic(same old behaviour) and then default Ingress Controller with Classic will get created. Setting LBType is not made mandatory so I have not set any default value.

File: pkg/asset/manifests/ingress_test.go (hunk 2 above)
Comment (miheer) [reply pkg/asset/manifests/ingress_test.go:104]: Actually that is not needed as for Classic LB setting LBType is not mandatory. The reason is LBType field comes under ProviderParameters which is not mandatory for Classic.

**Comment by miheer** [comment]
@sadasu PTAL

File: pkg/asset/manifests/ingress_test.go (hunk 2 above)
Comment (sadasu) [reply pkg/asset/manifests/ingress_test.go:104]: I understand that it is not mandatory. Shouldn't we have a test for all input values? It appears that we have one when it is not set, "NLB" but not one for when it is explicitly set to "Classic".

Comment (miheer) [reply pkg/asset/manifests/ingress_test.go:104]: Yes have added the test case for it as well. Please check line 106.

File: pkg/asset/manifests/ingress.go [hunk 3]
```diff
@@ -108,13 +108,24 @@ func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte,
...
-			Domain: fmt.Sprintf("apps.%s", config.ClusterDomain()),
-		},
+		Spec: configv1.IngressSpec{},
 		Status: configv1.IngressStatus{
 			DefaultPlacement: defaultPlacement,
 		},
 	}
+	obj.Spec.Domain = fmt.Sprintf("apps.%s", config.ClusterDomain())
+	if config.AWS != nil && len(config.AWS.LBType) != 0 {
+		obj.Spec = configv1.IngressSpec{
```
Comment (patrickdillon) [inline pkg/asset/manifests/ingress.go:118]: You have a bug here. This is causing you to lose the `obj.Spec.Domain`:

```shell
# cat c/manifests/cluster-ingress-02-config.yml 
apiVersion: config.openshift.io/v1
kind: Ingress
metadata:
  creationTimestamp: null
  name: cluster
spec:
  domain: ""
  loadbalancer:
    platform:
      aws:
        type: nlb
      type: AWS
status:
  defaultPlacement: Workers

```

Instead, you could just specify `LoadBalancer` without overwriting the whole spec:

```go
		obj.Spec.LoadBalancer = configv1.LoadBalancer{
				Platform: configv1.IngressPlatformSpec{
					AWS: &configv1.AWSIngressSpec{
						Type: config.AWS.LBType,
					},
					Type: configv1.AWSPlatformType,
				},
		}
```

File: pkg/asset/manifests/ingress.go [hunk 4]
```diff
@@ -108,13 +108,24 @@ func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte,
...
-		Spec: configv1.IngressSpec{
-			Domain: fmt.Sprintf("apps.%s", config.ClusterDomain()),
-		},
+		Spec: configv1.IngressSpec{},
 		Status: configv1.IngressStatus{
 			DefaultPlacement: defaultPlacement,
 		},
 	}
+	obj.Spec.Domain = fmt.Sprintf("apps.%s", config.ClusterDomain())
+	if config.AWS != nil && len(config.AWS.LBType) != 0 {
```
Comment (patrickdillon) [inline pkg/asset/manifests/ingress.go:117]: Typically we will use a switch statement because other platforms may need to make changes as well. https://github.com/openshift/installer/blob/master/pkg/asset/manifests/infrastructure.go#L95 is a good example of how to do this.

File: pkg/asset/manifests/ingress.go [hunk 5]
```diff
@@ -108,13 +109,24 @@ func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte,
...
 		},
-		Spec: configv1.IngressSpec{
-			Domain: fmt.Sprintf("apps.%s", config.ClusterDomain()),
-		},
+		Spec: configv1.IngressSpec{},
 		Status: configv1.IngressStatus{
 			DefaultPlacement: defaultPlacement,
 		},
 	}
+	obj.Spec.Domain = fmt.Sprintf("apps.%s", config.ClusterDomain())
```
Comment (r4f4) [inline pkg/asset/manifests/ingress.go:117]: Are these changes needed? What's wrong with the way `Spec.Domain` was being set?

File: pkg/asset/manifests/ingress.go [hunk 6]
```diff
@@ -108,13 +109,24 @@ func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte,
...
+	//if config.AWS != nil && len(config.AWS.LBType) != 0 {
```
Comment (r4f4) [inline pkg/asset/manifests/ingress.go:118]: Please, remove commented out code

File: pkg/types/aws/platform.go [hunk 7]
```diff
@@ -59,6 +62,20 @@ type Platform struct {
...
+	// Valid values are:
+	// * "Classic": A Classic Load Balancer that makes routing decisions at either
+	//   the transport layer (TCP/SSL) or the application layer (HTTP/HTTPS). See
+	//   the following for additional details:
+	//     https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#clb
+	// * "NLB": A Network Load Balancer that makes routing decisions at the
+	//   transport layer (TCP/SSL). See the following for additional details:
+	//     https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#nlb
+	// +optional
+	LBType configv1.AWSLBType `json:"lbType,omitempty"`
```
Comment (r4f4) [inline pkg/types/aws/platform.go:78]: Nit: this comment is adding weird spacing in the generated docs.

**Comment by r4f4** [comment]
`verify-codegen` is failing because you need to run `./hack/go-genmock.sh` and commit the result.

File: pkg/types/aws/platform.go (hunk 7 above)
Comment (miheer) [reply pkg/types/aws/platform.go:78]: @r4f4  does it look OK to you now ?

**Comment by candita** [comment]
CI env failure investigation: https://coreos.slack.com/archives/C68TNFWA2/p1665420718939409 and https://coreos.slack.com/archives/CBN38N3MW/p1665416502493189

**Comment by candita** [comment]
https://github.com/openshift/release/pull/32996 should make unit and verify-codegen tests optional.
/retest-required

**Comment by LorbusChris** [comment]
@miheer this PR needs a rebase

**Comment by candita** [comment]
Before, the only merge conflict was:
CONFLICT (rename/rename): Rename terraform/providers/random/vendor/github.com/dustinkirkland/golang-petname/LICENSE->vendor/cloud.google.com/go/compute/LICENSE in HEAD. Rename vendor/cloud.google.com/go/LICENSE->vendor/cloud.google.com/go/compute/LICENSE in https://github.com/openshift/installer/commit/b81fae1ed30691e42072c2b3961fc9b2a3d4a668

Now, there are also:
Auto-merging vendor/modules.txt
CONFLICT (content): Merge conflict in vendor/modules.txt
Auto-merging vendor/github.com/openshift/api/operator/v1/zz_generated.swagger_doc_generated.go
CONFLICT (content): Merge conflict in vendor/github.com/openshift/api/operator/v1/zz_generated.swagger_doc_generated.go
Auto-merging vendor/github.com/openshift/api/operator/v1/zz_generated.deepcopy.go
CONFLICT (content): Merge conflict in vendor/github.com/openshift/api/operator/v1/zz_generated.deepcopy.go
Auto-merging vendor/github.com/openshift/api/operator/v1/types_ingress.go
Auto-merging vendor/github.com/openshift/api/operator/v1/types_console.go
CONFLICT (content): Merge conflict in vendor/github.com/openshift/api/operator/v1/types_console.go
Auto-merging vendor/github.com/openshift/api/operator/v1/0000_70_dns-operator_00.crd.yaml
Auto-merging vendor/github.com/openshift/api/operator/v1/0000_70_console-operator.crd.yaml
CONFLICT (content): Merge conflict in vendor/github.com/openshift/api/operator/v1/0000_70_console-operator.crd.yaml
Auto-merging vendor/github.com/openshift/api/operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml
Auto-merging vendor/github.com/openshift/api/config/v1/zz_generated.swagger_doc_generated.go
CONFLICT (content): Merge conflict in vendor/github.com/openshift/api/config/v1/zz_generated.swagger_doc_generated.go
Auto-merging vendor/github.com/openshift/api/config/v1/zz_generated.deepcopy.go
CONFLICT (modify/delete): vendor/github.com/openshift/api/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml deleted in HEAD and modified in b81fae1ed30691e42072c2b3961fc9b2a3d4a668. Version b81fae1ed30691e42072c2b3961fc9b2a3d4a668 of vendor/github.com/openshift/api/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml left in tree.

**Comment by gcs278** [comment]
This PR has been open a while and now, there is no need to vendor anything because the current openshift-api version is: `20221004120407-c46852673d03`. 

I'm going to try to remove the vendoring commits, but not sure if I can modify this PR.

**Comment by gcs278** [comment]
I have opened https://github.com/openshift/installer/pull/6478 in lieu of this one so I could modify. I just removed the vendoring changes and rebased. 

PTAL
